tests/50federation/34room-backfill.pl

```diff
@@ -447,3 +447,113 @@
           });
       });
    };
+
+test "Outbound federation rejects backfill containing invalid JSON for events in room version 6",
+   requires => [ local_user_fixture(), $main::INBOUND_SERVER, federation_user_id_fixture() ],
+
+   do => sub {
+      my ( $user, $inbound_server, $creator_id ) = @_;
+
+      my $local_server_name = $inbound_server->server_name;
+      my $datastore         = $inbound_server->datastore;
+
+      my $room_alias = "#50fed-31backfill:$local_server_name";
+
+      my $room = $datastore->create_room(
+         creator      => $creator_id,
+         alias        => $room_alias,
+         room_version => "6",
+      );
+
+      # Create some past messages to backfill from
+      $room->create_and_insert_event(
+         type => "m.room.message",
+
+         sender  => $creator_id,
+         content => {
+            msgtype => "m.text",
+            body    => "Message here",
+            bad_val => 1.1,
+         },
+      );
+
+      Future->needs_all(
+         $inbound_server->await_request_backfill( $room->room_id )->then( sub {
+            my ( $req ) = @_;
+
+            # The helpfully-named 'v' parameter gives the "versions", i.e. the
+            # event IDs to start the backfill walk from. This can just be used
+            # in the 'start_at' list for $datastore->get_backfill_events.
+            # This would typically be an event ID the requesting server is
+            # aware exists but has not yet seen, such as one listed in a
+            # prev_events or auth_events list.
+            my $v     = $req->query_param( 'v' );
+
+            my $limit = $req->query_param( 'limit' );
+
+            my @events = $datastore->get_backfill_events(
+               start_at => [ $v ],
+               limit    => $limit,
+            );
+
+            log_if_fail "Responding with JSON", @events;
+
+            $req->respond_json( {
+               origin           => $inbound_server->server_name,
+               origin_server_ts => $inbound_server->time_ms,
+               pdus             => \@events,
+            } );
+
+            Future->done;
+         }),
+
+         do_request_json_for( $user,
+            method => "POST",
+            uri    => "/r0/join/$room_alias",
+
+            content => {},
+         )->then( sub {
+            my ( $body ) = @_;
+
+            my $room_id = $body->{room_id};
+
+            # It make take some time for the join to propagate, so we need to
+            # retry on failure.
```

Comment:
... or possibly `matrix_join_room_synced`.

As a general rule, tests that just `retry_until_success` are horrible to debug: they never fail outright, they just time out, without any clues as to why they were failing.